posts/2021/01/pytorch-tensoriterator-internals-update.md

```diff
@@ -0,0 +1,264 @@
+<!--
+.. title: PyTorch TensorIterator Internals (TensorIteratorConfig Update)
+.. slug: pytorch-tensoriterator-internals-update
+.. date: 2021-01-27 00:40:00 UTC-06:00
+.. author: Kurt Mohler
+.. tags: 
+.. category: 
+.. link: 
+.. description: 
+.. type: text
+-->
+
+For contributors to the PyTorch codebase, one of the most commonly encountered
+C++ classes is
+[`TensorIterator`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.h).
+`TensorIterator` offers a standardized way to iterate over elements of
+a tensor, automatically parallelizing operations, while abstracting device and 
+data type details.
+
+In April 2020, Sameer wrote a blog article discussing [PyTorch TensorIterator Internals](../../2020/04/pytorch-tensoriterator-internals.html). Recently,
+however, the interface has changed significantly. This post describes how to
+use the current interface as of January 2021. Much of the information from the
+previous article is directly copied here, but with updated API calls and some
+extra details.
+
+All of the code examples below can be compiled and run in [this GitHub
+repo](https://github.com/kurtamohler/pytorch-TensorIterator-examples).
+
+# Basics of TensorIterator and TensorIteratorConfig
+
+In order to create a `TensorIterator`, a `TensorIteratorConfig` must be created
+first. `TensorIteratorConfig` specifies the input and output tensors that will
+be iterated over, whether all tensors are expected to share the same data type
+and device, and a handful of other settings. After setting up the
+configuration, we can call `TensorIteratorConfig::build()` to obtain
+a `TensorIterator` that has the specified settings. `TensorIterator` is
+immutable, so once it is created, its configuration cannot be changed.
+
+In the following example, a tensor named `out` is configured as the output
+tensor and `a` and `b` are the input tensors. Calling `build` creates the 
+`TensorIterator` object from the specified configuration.
+
+``` cpp
+at::TensorIteratorConfig iter_config;
+iter_config
+  .add_output(out)
+  .add_input(a)
+  .add_input(b);
+
+auto iter = iter_config.build();
+```
+
+# Performing iterations
+
+Iterations using `TensorIterator` can be classified as point-wise iterations or
+reduction iterations. This plays a fundamental role in how iterations using
+`TensorIterator` are parallelized - point-wise iterations can be freely
+parallelized along any dimension and grain size while reduction operations have
+to be either parallelized along dimensions that you're not iterating over or by
+performing bisect and reduce operations along the dimension being iterated.
+Parallelization can also happen using vectorized operations. Parallelization
```

Comment:
Relevant for CUDA: normally you still parallelize over the reduction dimension, but then you will need synchronizations (atomics for example) to avoid race conditions.